### PR TITLE
Updated local development docker setup

### PR DESCRIPTION
**Pull request by b12f**
There are now separate containers for testing purposes, and the frontend container is a node-only image. Containers run
with a user that has the current user id and gid, making sure that any files that are installed by bundler or npm are
user-owned.

Selenium tests can be ran, but not all of them work yet.

### CODE REVIEW

File: docker/dev/backend/scripts/run-app [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+set -e
+set -u
+
+rm -rf "${OPENPROJECT_STORAGE_TMP__PATH:-tmp}/*"
+bundle exec rails server thin -b 0.0.0.0
```
Comment (crohr) [inline docker/dev/backend/scripts/run-app]: Why `thin` instead of `puma`?

**Review by crohr — COMMENTED** [review]
Looks good, although I wonder if it wouldn't be better with a base `docker-compose.dev.yml`, and an additional `docker-compose.test.yml` that overrides the various env variables and build targets as necessary?

Usually I have the wrapper (`bin/compose` in your case) take an env name as first argument so that it can automatically specify the additional compose files needed for docker-compose. This way you don't have to duplicate services just for the test environment.

I.e. 

```
# bin/compose
export COMPOSE_FILE="docker-compose.dev.yml"

env="$1" ; shift
if [ "$env" == "test" ]; then
  export COMPOSE_FILE="$COMPOSE_FILE:docker-compose.test.yml"
fi

exec docker-compose "$@"
```

I suppose you are familiar with https://docs.docker.com/compose/extends/#multiple-compose-files, but just in case.

File: docker/dev/backend/scripts/run-app (hunk 1 above)
Comment (b12f) [reply docker/dev/backend/scripts/run-app]: I took this from here https://github.com/opf/openproject/pull/8762/files#diff-ed0579942196f4cba4186b4e6e051eeeL16 and am too inexperienced with Ruby to have an opinion on which is better.

**Comment by b12f** [comment]
> Looks good, although I wonder if it wouldn't be better with a base `docker-compose.dev.yml`, and an additional `docker-compose.test.yml` that overrides the various env variables and build targets as necessary?

Would be possible, however the container names would still have to differ, otherwise docker will e.g. take old volumes when switching the overrides. So in that case why not just have the test containers in with the rest and be explicit in what you want to start (e.g. `./bin/compose up frontend backend` will not start the testing stuff)

> Usually I have the wrapper (`bin/compose` in your case) take an env name as first argument so that it can automatically specify the additional compose files needed for docker-compose. This way you don't have to duplicate services just for the test environment.

I get why you'd want this, I usually like to keep any docker-compose wrapping to a minimum because it quickly devolves into a mess where its really hard to do what you want. Personally, I just like to use the `docker-compose` command directly, and then have an `.env` in the base dir, but I felt like I perhaps shouldn't make too many changes to where files are located just yet.

In any case, I don't think we're going to get around duplicating services for tests. For the database, it just makes sense to start a different instance. For the backend/testing container, we need to install e.g. chromium and firefox (though we might be able to get around this at some point) which, if we'd be using the container for development too, would increase the difference between local development and any deployment setup.

File: docker/dev/backend/scripts/run-app (hunk 1 above)
Comment (crohr) [inline docker/dev/backend/scripts/run-app]:
```suggestion
exec bundle exec rails server -u puma -b 0.0.0.0
```

**Review by crohr — APPROVED** [review]
All fine by me. In any case I don't think this would break anyone's workflow since I don't think @ulferts or @machisuji use the compose-based env?

I.e. don't hesitate to make wide changes :) Just add a section in the README maybe with the minimum commands required to launch the dev (and possible test) env(s)?

**Review by machisuji — APPROVED** [review]
:+1:

**Pull request merged**